crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+            Self(MIN_DECIMAL_SCALE)
+        }
+
+        pub const fn max() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+
+        pub fn apply(&self, val: &mut Decimal) {
+            val.rescale(self.0)
+        }
+    }
+
+    impl TryFrom<u32> for Scale {
+        type Error = String;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                x if (MIN_DECIMAL_SCALE..=MAX_DECIMAL_SCALE).contains(&x) => Ok(Scale(x)),
+                _ => Err(format!(
+                "Numeric scale {value} must be between {MIN_DECIMAL_SCALE} and {MAX_DECIMAL_SCALE}"
+            )),
+            }
+        }
+    }
+
+    impl Display for Scale {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Scale {
+        fn default() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+    }
+}
+
+use inner_types::*;
+
+/// A wrapper type around [`rust_decimal::Decimal`].
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, SizeOf)]
+#[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "with-serde", serde(transparent))]
+#[derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
+pub struct SQLDecimal {
+    inner: Decimal,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_precision: Precision,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_scale: Scale,
+}
+
+impl FromStr for SQLDecimal {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let inner = Decimal::from_str(s).map_err(|e| e.to_string())?;
+
+        Ok(Self {
+            inner,
+            max_precision: Precision::max(),
+            max_scale: Scale::try_from(inner.scale())?,
+        })
+    }
+}
+
+impl TryFrom<Decimal> for SQLDecimal {
+    type Error = String;
+
+    fn try_from(mut inner: Decimal) -> Result<Self, Self::Error> {
+        let scale = Scale::max();
+        scale.apply(&mut inner);
+
+        let precision = Precision::max();
+
+        Ok(Self {
+            inner,
+            max_precision: precision,
+            max_scale: scale,
+        })
+    }
+}
+
+impl SQLDecimal {
+    pub fn new(precision: u32, scale: u32) -> Self {
+        let inner = Default::default();
+
+        let scale = Scale::try_from(scale).unwrap();
+        let precision = Precision::try_from(precision).unwrap();
+
+        Self {
+            inner,
+            max_precision: precision,
+            max_scale: scale,
+        }
+    }
+
+    pub fn from_str_with_precision_scale(
+        s: &str,
+        _precision: u32,
+        scale: u32,
+    ) -> Result<Self, String> {
+        let mut inner = Decimal::from_str(s).map_err(|e| e.to_string())?;
+
+        // TODO: use the precision value from the parameter
+        let max_precision = Precision::max();
```

Comment:
why aren't you?

Reply:
I wasn't using it because it broke the Postgres tests as the precision of expected output was different than the precision of calculated output. 

But now, our implementation will be different from Postgres so I will use it.